doc/introduction/mid_circuit_measurements.rst

```diff
@@ -0,0 +1,530 @@
+
+.. _tree_traversal:
+
+Tree-traversal algorithm
+************************
+
+Dynamic circuit execution is akin to traversing a binary tree where each mid-circuit measurement
+corresponds to a node and groups of gates between them correspond to edges.
+The :func:`~.pennylane.dynamic_one_shot` approach above picks a branch of the tree at random
+and simulates it from beginning to end.
+This is wasteful in many cases; the same branch is simulated many times
+when there are more shots than branches for example, and shared information between branches
+is not reused.
+The tree-traversal algorithm does away with such redundancy while retaining the
+exponential gains in memory of the one-shot approach compared with the deferred
+measurement principle, among other advantages.
+
+The algorithm cuts an :math:`n_{MCM}` circuit into :math:`n_{MCM}+1`
+circuit segments. Each segment can be executed on either the 0- or 1-branch,
+which gives rise to a binary tree with :math:`2^{n_{MCM}}` leaves. Terminal
+measurements are obtained at the leaves, and propagated and combined back up at each
+node up the tree. The tree is explored using a depth-first pattern. The tree-traversal
+method improves on :func:`~.pennylane.dynamic_one_shot` by collecting all samples at a
+node or leaf at once. Neglecting overheads, simulating all branches requires the same
+amount of computations as :func:`~. pennylane.defer_measurements`, but without the
+:math:`O(2^{n_{MCM}})` memory requirement. To save time, a copy of the state vector
+is made at every node, or mid-circuit measurement, requiring :math:`n_{MCM}+1` state
+vectors, an exponential improvement compared to :func:`~.pennylane.defer_measurements`.
+Since the counts of many nodes come out to be zero for shot-based simulations,
+it is often possible to ignore entire sub-trees, thereby reducing the computational
+cost.
+
+To summarize, this algorithm gives us the best of both worlds. In the limit of few
+shots and/or many mid-circuit measurements, it is as fast as the naive shot-by-shot
+implementation of ``dynamic_one_shot`` because few sub-trees are explored.
+In the limit of many shots and/or few mid-circuit measurements, it is
+equal to or faster than the deferred measurement algorithm (albeit with more
+overheads in practice) because each tree edge is visited at most once, all while
+exponentially reducing the memory requirements.
+
+The tree-traversal algorithm is not a transform. Its usage is therefore specified
+by passing an ``mcm_method`` option to a QNode (see section
+:ref:`"Configuring mid-circuit measurements" <mcm_config>`). For example,
+
+.. code-block:: python
+
+    @qml.qnode(dev, mcm_method="tree-traversal")
+    def my_qnode(x, y):
+        (...)
+
+.. warning::
+
+    The tree-traversal algorithm is only implemented for the
+    :class:`~.pennylane.devices.DefaultQubit` device.
+
+.. _mcm_config:
+
+Configuring mid-circuit measurements
+************************************
```

Comment:
Just a quick comment, but I find this section quite difficult to read, as it is _super_ dense with a lot of `inline code`.

In addition, it is mostly documenting two QNode keyword arguments. Would it make more sense to simply _summarize_ the available configuration details, but link to the QNode docstring for more details?

This way:

- the details are close to the code, and easier to maintain
- readers who end up on the QNode docstring directly will get all the information they need there, without needing to look for this page

Reply:
I cut down this section a good bit. Let me know what you think!

Regarding linking to the docs, we currently do not document many features/aspects of QNodes in the docstring. There is an extensive parameter broadcasting toggle section, but otherwise it is very minimal.
We could work on this, but I would consider that a separate task.